### ISSUE TITLE
`if` statement in loop changes type of variable outside the `if`

### ISSUE THREAD

**Issue by andy-ms** [microsoft/TypeScript#19955]
**TypeScript Version:**  2.7.0-dev.20171112

**Code**

```ts
type U = { kind: "A", xs?: number[] } | { kind: "B", xs?: number[] };
function f(u: U) {
	if (u.xs) {
		for (const b of u.xs) {
			if (u.kind === "A") {}
		}
	}
}
```

**Expected behavior:**

No error.

**Actual behavior:**

`src/a.ts(4,19): error TS2532: Object is possibly 'undefined'.`

**Comment by ahejlsberg** [comment]
Duplicate of #18840.

**Comment by ahejlsberg** [comment]
Actually, I'll flip it back to a bug since it is not *exactly* the same as #18840. However the root cause is similar: A narrowing by a discriminant property (such as `u.kind === "A"`) resets any previous narrowing in effect for the variable. Here, the back edge of the for loop causes the narrowing of `u.xs` to disappear. This is unfortunately not a simple thing to fix.

**Comment by andy-ms** [comment]
Another case:
```ts
function containsItself(arr: {}): boolean {
    if (!Array.isArray(arr)) {
        return false;
    }
    for (const x of arr) {
        if (x === arr) {
            return true;
        }
    }
    return false;
}
```

```
src/a.ts(5,16): error TS7022: 'x' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
```

**Comment by andy-ms** [comment]
A possibly related error:
```ts
function f(x: A | B): void {
	if (typeof x.a === "number") {
		if (x.kind === "A") { } // No error if this is removed
		const n: number = x.a; // Error
	}
}
```

**Comment by andy-ms** [comment]
Another case:
```ts
function f(a: number | undefined) {
    if (!a) return;
    for (const b of [a]) {
        const c = b;
        if (c === a) {}
    }
}
```

```
src/a.ts:3:16 - error TS7022: 'b' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.

3     for (const b of [a]) {
                 ~


src/a.ts:4:15 - error TS7022: 'c' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.

4         const c = b;
                ~
```

**Comment by andy-ms** [comment]
Another case from #26673:
```ts
interface A {
	isA: true;
	m?(): void;
}

interface B {
	isA: false;
	m?(): void;
}

class User {
	p!: A | B;

	go(): void {
		if (this.p.m) {
			if (this.p.isA) {}
			this.p.m(); // No error (good)
		}

		const p = this.p;
		if (p.m) {
			if (p.isA) {}
			p.m(); // Error (bad)
		}
	}
}
```

**Comment by jack-williams** [comment]
This looks to be fixed by #32695.

**Comment by sandersn** [comment]
Indeed it is. Thanks for noticing!
